### PR TITLE
[L] store jwt in browser cookie.

### PR DESCRIPTION
**Pull request by subhankarb**
we use JWT and store them in localstorage. Since these are not sent as
 part of the request to the server there is no way for the server
 to know if user is authenticated.

 We will store jwt inside cookie, so that we know user is authenticated ?

   * Added set cookie in callback handler
   * Added cookie unset method in logout
   * Added get_user_from_cookie method to extract user from cookie.
   * Update all site urls. If JWT is invalid all will redirect to /logout page.
   * Update base.html navbar login/logout button based on user variable from backend.
     * Removed all relative js code.
   * Removed encoded_token storing js method.

   Additional Fixes
   * Update JWT expiration_hour to 24.
   * Updated search swagger doc.

Resolves : #292

### CODE REVIEW

**Comment by rufuspollock** [comment]
@subhankarb does the 24h expiration mean the user will have to login again after 24h?

**Comment by rufuspollock** [comment]
@subhankarb suggest this could have been two commits:

* commit for JWT changes
* commit for template changes

Would have made it much easier to review and would have reflected issue structure.

File: app/auth/models.py [hunk 1]
```diff
@@ -10,8 +10,9 @@
 
 
 class JWT(object):
```
Comment (rufuspollock) [inline app/auth/models.py]: Does this change expiration for both API and browser. I think we probably want different times for the two locations. For the API tokens should be very short as the client can renew. For browser we probably want long tokens (a month or a year).

File: app/search/controllers.py [hunk 2]
```diff
@@ -24,8 +24,8 @@ def search_packages():
             - in: query
               name: q
               type: string
-              required: true
-              description: search query string e.g. q=query publisher=pub
+              required: false
```
Comment (rufuspollock) [inline app/search/controllers.py]: This seems an unrelated change. Why is it in this PR? What does it actually relate to?

Do you do `git diff --cached` before committing and check what will be included?

File: app/site/controllers.py [hunk 3]
```diff
@@ -31,23 +31,14 @@ def index():
...
-                    return redirect(request.headers['Host'] + '/logout')
-                user = User().get_userinfo_by_id(payload['user'])
-                if user:
-                    return render_template("dashboard.html", user=user,
-                                           title='Dashboard'), 200
-                return redirect(request.headers['Host'] + '/logout')
-        return render_template("index.html", title='Home'), 200
-    except Exception:
-        return redirect(url_for('.logout'))
+    user, exception = get_user_from_cookie()
```
Comment (rufuspollock) [inline app/site/controllers.py]: In terms of code logic I think:

* `if exception` should come first: handle the error condition immediately
* It would read better as a chained if statement so logic was super clear

```
if exception:
  return logout
elif user:
  return dashboard
else:
  return index.html

File: app/site/controllers.py [hunk 4]
```diff
@@ -31,23 +31,14 @@ def index():
...
+    if user:
+        return render_template("dashboard.html",
+                               user=user,
+                               title='Dashboard'), 200
+    if exception:
+        return redirect(request.headers['Host'] + '/logout')
```
Comment (rufuspollock) [inline app/site/controllers.py]: Why is this not simply:

return redirect(url_for(logout))

File: app/site/controllers.py [hunk 5]
```diff
@@ -61,7 +52,9 @@ def logout():
       302:
         description: Load the Home Page
     """
-    return render_template("logout.html", title='Logout'), 200
+    resp = make_response(render_template("logout.html", title='Logout'), 200)
+    resp.set_cookie('jwt', '', expires=0)
```
Comment (rufuspollock) [inline app/site/controllers.py]: Can't you just delete the cookie?

File: app/site/controllers.py [hunk 6]
```diff
@@ -87,6 +80,10 @@ def datapackage_show(publisher, package):
       200:
         description: Succesfuly loaded
```
Comment (rufuspollock) [inline app/site/controllers.py]: Typo in `Successfully`: please english spell check comments ;-)

File: app/site/controllers.py [hunk 7]
```diff
@@ -109,7 +106,8 @@ def datapackage_show(publisher, package):
         .split('\n\n')[0].replace(' \n', '') \
         .replace('\n', ' ').replace('/^ /', '')
 
-    return render_template("dataset.html", dataset=dataset,
+    return render_template("dataset.html", user=user,
```
Comment (rufuspollock) [inline app/site/controllers.py]: User (or better something a bit more descriptive like `current_user`) should not be set in every method but injected globally into the tmeplates see e.g. http://stackoverflow.com/questions/26498689/flask-jinja-pass-data-to-a-base-template-all-templates

File: app/site/controllers.py [hunk 8]
```diff
@@ -87,6 +80,10 @@ def datapackage_show(publisher, package):
       200:
         description: Succesfuly loaded
     """
+    user, exception = get_user_from_cookie()
```
Comment (rufuspollock) [inline app/site/controllers.py]: We should not be repeating the same code over and over again in different locations! DRY!

If what you want here is to set the logged user (if there is one) before every request this is a very common pattern and should be done in one global location. Please do a quick google or see http://stackoverflow.com/questions/19531721/flask-g-user-and-before-request

File: app/templates/base.html [hunk 9]
```diff
@@ -35,7 +35,17 @@
...
 
           <ul class="nav navbar-nav navbar-right">
-            <li><div id="nav-bar-login"></div></li>
+            {% if not user %}
+              <li>
+                <a href="/api/auth/login">Sign In With <span class="fa fa-github fa-lg"></span></a>
+              </li>
+            {% else %}
+              <li>
+                <button class="btn btn-small btn-outline-success p-0">
```
Comment (rufuspollock) [inline app/templates/base.html]: Why is this a button but the sign in is <a>. I recommend <a> by default -- it's a link here!

File: app/site/controllers.py (hunk 4 above)
Comment (subhankarb) [inline app/site/controllers.py]: @rufuspollock redirect(url_for(logout)) takes `API GATEWAY` url. As we are redirecting from dreamhost to apigateway.
So redirect(url_for(logout)) would take `aws.dpr-api-{hash}/stage`

**Comment by subhankarb** [comment]
does the 24h expiration mean the user will have to login again after 24h?
Yes

File: app/auth/models.py (hunk 1 above)
Comment (subhankarb) [inline app/auth/models.py]: @rufuspollock This can be done. As the flows for getting toke for site and dpmpy are  different. 
The site gets the token from API `{api}/auth/callback` and client gets the token from API `{api}/auth/token`. We can set different expiry time.

File: app/search/controllers.py (hunk 2 above)
Comment (subhankarb) [inline app/search/controllers.py]: @rufuspollock I agree this is completely unrelated. That is why is also mentioned in commit message about this change. But it should not be here.

File: app/__init__.py [hunk 10]
```diff
@@ -108,6 +110,19 @@ def get_github_oauth_token():
...
+
+    @app.before_request
+    def get_user_from_cookie():
+        token = request.cookies.get('jwt')
+        g.current_user, g.jwt_exception = None, None
+        if token:
+            try:
+                payload = JWT(app.config['API_KEY']).decode(token)
+                g.current_user = User().get_userinfo_by_id(payload['user'])
+            except Exception as e:
```
Comment (rufuspollock) [inline app/__init__.py]: Why are we catching exceptions here?

File: app/auth/controllers.py [hunk 11]
```diff
@@ -56,7 +56,8 @@ def callback_handling():
         user = User().create_or_update_user_from_callback(user_info.data)
         jwt_helper = JWT(app.config['API_KEY'], user.id)
         session.pop('github_token', None)
-        resp = make_response(render_template("dashboard.html", user=user,
+        g.current_user = user
```
Comment (rufuspollock) [inline app/auth/controllers.py]: Why are you setting g.current_user here? Surely you have just set the cookie so on the next page request they will be logged in fine.

File: app/auth/controllers.py [hunk 12]
```diff
@@ -56,7 +56,8 @@ def callback_handling():
...
+        resp = make_response(render_template("dashboard.html",
```
Comment (rufuspollock) [inline app/auth/controllers.py]: You don't want to do the render here - you want to redirect to dashboard page (which is `/` for now)

File: app/site/controllers.py [hunk 13]
```diff
@@ -27,17 +27,16 @@ def index():
...
       404:
-        description: Publiser does not exist
+        description: Publisher does not exist
       200:
-        description: Succesfuly loaded home page
+        description: Successfully loaded home page
     """
-    user, exception = get_user_from_cookie()
-    if user:
+    if g.jwt_exception:
```
Comment (rufuspollock) [inline app/site/controllers.py]: We don't want to check for g.jwt_exception on every request. This should be somewhere common (e.g. before request handler or something like that -- i'm not quite sure how flask supports this).

**More important**: why are we handling jwt_exception and redirecting to logout? Why can't we just proceed as normal -- the user just won't be logged in!

File: app/site/controllers.py [hunk 14]
```diff
@@ -27,17 +27,16 @@ def index():
...
+        return redirect(request.headers['Host'] + '/logout')
```
Comment (rufuspollock) [inline app/site/controllers.py]: Why aren't we using http://flask.pocoo.org/docs/0.12/api/#flask.url_for - and why are we using request.headers stuff?

File: app/site/controllers.py [hunk 15]
```diff
@@ -52,6 +51,7 @@ def logout():
       302:
```
Comment (rufuspollock) [inline app/site/controllers.py]: Extra comment: description says "loads home page if user already login" - is that correct? That seems weird. (Also note incorrect english grammar: if user already log**ged** in"

File: app/site/controllers.py [hunk 16]
```diff
@@ -52,6 +51,7 @@ def logout():
...
         description: Load the Home Page
     """
+    g.current_user = None
```
Comment (rufuspollock) [inline app/site/controllers.py]: Why do we need to set this here? Surely once we've expired the cookie it won't get set on the next request.

File: app/site/controllers.py [hunk 17]
```diff
@@ -52,6 +51,7 @@ def logout():
...
     resp = make_response(render_template("logout.html", title='Logout'), 200)
```
Comment (rufuspollock) [inline app/site/controllers.py]: It would make more sense to set a flash message and redirect them to the home page.

File: app/site/controllers.py [hunk 18]
```diff
@@ -52,6 +51,7 @@ def logout():
...
     resp.set_cookie('jwt', '', expires=0)
```
Comment (rufuspollock) [inline app/site/controllers.py]: Why aren't we deleting the cookie?

Comment (rufuspollock) [inline app/site/controllers.py]: Also, I guess there is no way to expire the JWT (that's how JWT's work right ...?)

File: app/site/controllers.py [hunk 19]
```diff
@@ -78,10 +78,9 @@ def datapackage_show(publisher, package):
       404:
         description: Datapackage does not exist
       200:
-        description: Succesfuly loaded
+        description: Successfully loaded
     """
-    user, exception = get_user_from_cookie()
-    if exception:
+    if g.jwt_exception:
```
Comment (rufuspollock) [inline app/site/controllers.py]: As per above - i won't keep commenting on these ones ;-)

File: app/__init__.py (hunk 10 above)
Comment (subhankarb) [inline app/__init__.py]: @rufuspollock If there is a problem with jwt decode the exception will throw a 500 error.
So that is why trying to catch if any decode exception e.g token expired is there and show login button in navbar. 
In this process user can try to login again and a fresh jwt token will be assigned to the user.
And I tried to capture all exception as i do not want to have any failure here.
But `g.jwt_exception = e` is completely unnecessary

File: app/site/controllers.py (hunk 15 above)
Comment (subhankarb) [inline app/site/controllers.py]: FIXED.
